internal/querywatcher/query_watcher.go

```diff
@@ -172,20 +174,50 @@ func (w *QueryManager) updateQueryCache(queryFingerprint string, event dstore.Wa
 	}
 }
 
-// notifyClient notifies all clients watching a query about the new result.
 func (w *QueryManager) notifyClients(query *DSQLQuery, clients *sync.Map, queryResult *[]dstore.DSQLQueryResultRow) {
 	encodedResult := clientio.Encode(CreatePushResponse(query, queryResult), false)
-	// Send the result to the client
+
 	clients.Range(func(clientKey, _ interface{}) bool {
 		clientFD := clientKey.(int)
-		_, err := syscall.Write(clientFD, encodedResult)
-		if err != nil {
-			w.removeWatcher(query, clientFD)
-		}
+		// We use a retry mechanism here as the client's socket may be temporarily unavailable for writes due to the
+		// high number of writes that are possible in qwatch. Without this mechanism, the client may be removed from the
+		// watchlist prematurely.
+		// TODO:
+		//  1. Replace with thread pool to prevent launching an unbounded number of goroutines.
+		//  2. Each client's writes should be sent in a serialized manner, maybe a per-client queue should be maintained
+		//   here. A single queue-per-client is also helpful when the client file descriptor is closed and the queue can
+		//   just be destroyed.
+		go w.sendWithRetry(query, clientFD, encodedResult)
 		return true
 	})
 }
 
+// sendWithRetry writes data to a client file descriptor with retries. It writes with an exponential backoff.
+func (w *QueryManager) sendWithRetry(query *DSQLQuery, clientFD int, data []byte) {
+	maxRetries := 20
```

Comment:
Do we need max 20 retries? We can reduce the number. Also can we add random jitter to it? Just a thought so that all the clients won't collide. 

Thanks for the change.

Reply:
Just a random number, this can be tuned based on more fine-grained testing later on